Exceptions when mpHealth-1.0 and sessionCache-1.0 are enabled

When both mpHealth-1.0 and sessionCache-1.0 are enabled, several exceptions occur. See attached log.

[messages.log_mp_sc.txt](https://github.com/OpenLiberty/open-liberty/files/1838678/messages.log_mp_sc.txt)

@dacleyra 
It looks like this might be an issue with newer Java versions. I tried the app on the following JDKs:
1.8.0_71 (IBM): No exceptions
1.8.0_151 (IBM): No exceptions
1.8.0_161 (IBM): An unexpected exception occurred when trying to retrieve the session context
java.security.ProviderException: setSeed() failed
    at sun.security.provider.NativePRNG$RandomIO.implSetSeed(NativePRNG.java:472)
    at sun.security.provider.NativePRNG$RandomIO.access$300(NativePRNG.java:331)
    at sun.security.provider.NativePRNG.engineSetSeed(NativePRNG.java:214)
    at java.security.SecureRandom.setSeed(SecureRandom.java:464)
....
Caused by: java.io.IOException: Operation not permitted
    at java.io.FileOutputStream.writeBytes(Native Method)
    at java.io.FileOutputStream.write(FileOutputStream.java:313)
    at sun.security.provider.NativePRNG$RandomIO.implSetSeed(NativePRNG.java:470)

1.8.0_161 (Oracle): SRVE8059E: An unexpected exception occurred when trying to retrieve the session context
com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.ClassNotFoundException: javax.cache.configuration.FactoryBuilder$SingletonFactory
....
Caused by: java.lang.ClassNotFoundException: javax.cache.configuration.FactoryBuilder$SingletonFactory

--- a/dev/com.ibm.websphere.javaee.jcache.1.1/bnd.bnd
+++ b/dev/com.ibm.websphere.javaee.jcache.1.1/bnd.bnd
@@ -14,15 +14,7 @@ bVersion=1.0
 Bundle-SymbolicName: com.ibm.websphere.javaee.jcache.1.1
 
 Export-Package: \
-   javax.cache;version="1.1", \
-   javax.cache.annotation;version="1.1", \
-   javax.cache.configuration;version="1.1", \
-   javax.cache.event;version="1.1", \
-   javax.cache.expiry;version="1.1", \
-   javax.cache.integration;version="1.1", \
-   javax.cache.management;version="1.1", \
-   javax.cache.processor;version="1.1", \
-   javax.cache.spi;version="1.1"
+   javax.cache.*;version="1.1";thread-context=true
 
 Import-Package: \
    javax.enterprise.util;resolution:=optional

--- a/dev/com.ibm.ws.session.cache_fat/build.gradle
+++ b/dev/com.ibm.ws.session.cache_fat/build.gradle
@@ -11,11 +11,13 @@
 
 configurations {
   hazelcastJCache
+  hazelcastClient
 }
 
 // Define G:A:V coordinates of each dependency
 dependencies {
   hazelcastJCache 'com.hazelcast:hazelcast:3.9.2'
+  hazelcastClient 'com.hazelcast:hazelcast-client:3.9.2'
 }
 
 task addHazelcastJCacheProviderToSharedDir(type: Copy) {
@@ -24,7 +26,14 @@ task addHazelcastJCacheProviderToSharedDir(type: Copy) {
   rename 'hazelcast-.*.jar', 'hazelcast.jar'
 }
 
+task addHazelcastClientToSharedDir(type: Copy) {
+  from configurations.hazelcastClient
+  into new File(autoFvtDir, 'publish/shared/resources/hazelcast')
+  rename 'hazelcast-client-*.jar', 'hazelcast-client.jar'
+}
+
 addRequiredLibraries {
   dependsOn addDerby
   dependsOn addHazelcastJCacheProviderToSharedDir
+  dependsOn addHazelcastClientToSharedDir
 }

--- a/dev/com.ibm.ws.session.cache_fat/fat/src/com/ibm/ws/session/cache/fat/FATSuite.java
+++ b/dev/com.ibm.ws.session.cache_fat/fat/src/com/ibm/ws/session/cache/fat/FATSuite.java
@@ -29,7 +29,8 @@ import componenttest.topology.utils.HttpUtils;
                 SessionCacheOneServerTest.class,
                 SessionCacheTwoServerTest.class,
                 SessionCacheTimeoutTest.class,
-                SessionCacheTwoServerTimeoutTest.class
+                SessionCacheTwoServerTimeoutTest.class,
+                HazelcastClientTest.class
 })
 
 public class FATSuite {

--- a/dev/com.ibm.ws.session.cache_fat/fat/src/com/ibm/ws/session/cache/fat/HazelcastClientTest.java
+++ b/dev/com.ibm.ws.session.cache_fat/fat/src/com/ibm/ws/session/cache/fat/HazelcastClientTest.java
@@ -1,0 +1,80 @@
+/*******************************************************************************
+ * Copyright (c) 2018 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.session.cache.fat;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.UUID;
+
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import componenttest.annotation.Server;
+import componenttest.custom.junit.runner.FATRunner;
+import componenttest.custom.junit.runner.Mode;
+import componenttest.custom.junit.runner.Mode.TestMode;
+import componenttest.topology.impl.LibertyServer;
+import componenttest.topology.utils.FATServletClient;
+
+@RunWith(FATRunner.class)
+public class HazelcastClientTest extends FATServletClient {
+
+    @Server("sessionCacheServerA")
+    public static LibertyServer serverA;
+
+    @Server("sessionCacheServer-mphealth")
+    public static LibertyServer serverB;
+
+    @BeforeClass
+    public static void setUp() throws Exception {
+        serverB.useSecondaryHTTPPort();
+
+        String configLocation = new File(serverB.getUserDir() + "/shared/resources/hazelcast/hazelcast-client-localhost-only.xml").getAbsolutePath();
+        String rand = UUID.randomUUID().toString();
+        serverA.setJvmOptions(Arrays.asList("-Dhazelcast.group.name=" + rand));
+        serverB.setJvmOptions(Arrays.asList("-Dhazelcast.group.name=" + rand,
+                                            "-Dhazelcast.config=" + configLocation));
+    }
+
+    @AfterClass
+    public static void cleanup() throws Exception {
+        // Servers should be stopped at this point, but just in case a test fails
+        try {
+            if (serverB.isStarted())
+                serverB.stopServer();
+        } finally {
+            if (serverA.isStarted())
+                serverA.stopServer();
+        }
+    }
+
+    /**
+     * When a hazelcast client restarts, it drives down a code path where Hazelcast loads JCache
+     * API using the TCCL. When this code path is taken from a WAB (OSGi app), the JCache API
+     * packages need to be exposed to the WABs bundle TCCL. In this test serverB enabled mpHealth-1.0
+     * which uses a WAB to deploy the /health/ web endpoint.
+     */
+    @Test
+    @Mode(TestMode.FULL)
+    public void testWABClientRestart() throws Exception {
+        serverA.startServer(testName.getMethodName() + ".log");
+        serverB.startServer(testName.getMethodName() + "-before.log");
+
+        // Triggering a server restart of the hazelcast client will
+        serverB.stopServer();
+        serverB.startServer(testName.getMethodName() + "-after.log");
+
+        serverB.stopServer();
+        serverA.stopServer();
+    }
+}

--- a/dev/com.ibm.ws.session.cache_fat/publish/servers/sessionCacheServer-mphealth/bootstrap.properties
+++ b/dev/com.ibm.ws.session.cache_fat/publish/servers/sessionCacheServer-mphealth/bootstrap.properties
@@ -1,0 +1,12 @@
+###############################################################################
+# Copyright (c) 2018 IBM Corporation and others.
+# All rights reserved. This program and the accompanying materials
+# are made available under the terms of the Eclipse Public License v1.0
+# which accompanies this distribution, and is available at
+# http://www.eclipse.org/legal/epl-v10.html
+#
+# Contributors:
+#     IBM Corporation - initial API and implementation
+###############################################################################
+bootstrap.include=../testports.properties
+com.ibm.ws.logging.trace.specification=*=info:com.ibm.ws.session.store.cache=all

--- a/dev/com.ibm.ws.session.cache_fat/publish/servers/sessionCacheServer-mphealth/server.xml
+++ b/dev/com.ibm.ws.session.cache_fat/publish/servers/sessionCacheServer-mphealth/server.xml
@@ -1,0 +1,35 @@
+<server>
+
+    <featureManager>
+        <feature>cdi-2.0</feature>
+        <feature>servlet-4.0</feature>
+        <feature>componenttest-1.0</feature>
+        <feature>sessionCache-1.0</feature>
+        <feature>timedexit-1.0</feature>
+        <!-- MP Health is enabled to verify that sessionCache works with a feature that provides a web endpoint via WAB -->
+        <feature>mpHealth-1.0</feature>
+    </featureManager>
+    
+    <include location="../fatTestCommon.xml"/>
+
+    <!-- This server runs at the same time as sessionCacheServerA, so use a different set of ports -->    
+    <httpEndpoint id="defaultHttpEndpoint"
+                  httpPort="${bvt.prop.HTTP_secondary}"
+                  httpsPort="${bvt.prop.HTTP_secondary.secure}"/>
+    
+	<httpSessionCache libraryRef="HazelcastLib">
+	  <properties hazelcast.config.location="file:${shared.resource.dir}/hazelcast/hazelcast-client-localhost-only.xml"/>
+	</httpSessionCache>
+	
+	<library id="HazelcastLib">
+	  <fileset dir="${shared.resource.dir}/hazelcast" includes="*.jar"/>
+	</library>
+    
+    <!-- Needed for missing doPriv in JCache 1.1 API (see https://github.com/jsr107/jsr107spec/issues/398) -->
+    <javaPermission className="java.util.PropertyPermission" actions="read,write" name="*"/>
+    <!-- Needed for hazelcast JVM shutdown hook -->
+    <javaPermission className="java.lang.RuntimePermission" name="accessClassInPackage.sun.net.www.protocol.wsjar"/>
+    
+    <javaPermission codebase="${shared.resource.dir}/hazelcast/hazelcast.jar" className="java.security.AllPermission"/>
+    <javaPermission codebase="${shared.resource.dir}/hazelcast/hazelcast-client.jar" className="java.security.AllPermission"/>
+</server>

--- a/dev/com.ibm.ws.session.cache_fat/publish/shared/resources/hazelcast/hazelcast-client-localhost-only.xml
+++ b/dev/com.ibm.ws.session.cache_fat/publish/shared/resources/hazelcast/hazelcast-client-localhost-only.xml
@@ -1,0 +1,8 @@
+<hazelcast-client xsi:schemaLocation="http://www.hazelcast.com/schema/client-config hazelcast-client-config-3.9.xsd"
+           xmlns="http://www.hazelcast.com/schema/client-config"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+    <group>
+        <name>${hazelcast.group.name}</name>
+        <password>dev-pass</password>
+    </group>
+</hazelcast-client>

--- a/dev/com.ibm.ws.session.store/src/com/ibm/ws/session/store/common/BackedStore.java
+++ b/dev/com.ibm.ws.session.store/src/com/ibm/ws/session/store/common/BackedStore.java
@@ -20,6 +20,7 @@ import com.ibm.ws.session.SessionManagerConfig;
 import com.ibm.ws.session.SessionStoreService;
 import com.ibm.ws.session.store.common.internal.LoggingUtil;
 import com.ibm.ws.session.store.memory.MemoryStore;
+import com.ibm.wsspi.kernel.service.utils.FrameworkState;
 import com.ibm.wsspi.session.ILoader;
 import com.ibm.wsspi.session.ISession;
 import com.ibm.wsspi.session.IStoreCallback;
@@ -125,7 +126,13 @@ public abstract class BackedStore extends MemoryStore {
      */
     public void runInvalidation() {
         ((BackedHashMap) _sessions).cleanUpCache(System.currentTimeMillis());
-        ((BackedHashMap) _sessions).performInvalidation();
+        if(!FrameworkState.isStopping()) {
+            ((BackedHashMap) _sessions).performInvalidation();
+        } else {
+            if (com.ibm.websphere.ras.TraceComponent.isAnyTracingEnabled() && LoggingUtil.SESSION_LOGGER_WAS.isLoggable(Level.FINE)) {
+                LoggingUtil.SESSION_LOGGER_WAS.logp(Level.FINE, methodClassName, "runInvalidation", "Skipping background session invalidation because server is shutting down.");
+            }
+        }
         /*
          * if (_isApplicationSessionStore) {
          * ((BackedHashMap)_sessions).performAppSessionInvalidation();

